src/laser_core/migration.py

```diff
@@ -0,0 +1,394 @@
+    a (float): Exponent for the population of the origin.
+    b (float): Exponent parameter for populations in the gravity model.
+    c (float): Exponent parameter for distances in the gravity model.
+    delta (float): Exponent parameter for the competing destinations adjustment.
+    **params: Additional parameters to be passed to the gravity model.
+    Returns:
+    numpy.ndarray: Adjusted network matrix based on the competing destinations model.
+    """
+
+    # Sanity checks
+    shared_sanity_checks(pops, distances, a=a, k=k, b=b, c=c, delta=delta)
+
+    network = gravity(pops, distances, k=k, a=a, b=b, c=c, **params)
+    # Construct the p_j^b / d_jk^c matrix, inside the sum
+    distances1 = distances.copy()
+    np.fill_diagonal(distances1, 1)  # Prevent division by zero in `distances ** (-1 * c)`
+    competition_matrix = pops**b * distances1 ** (-1 * c)
+
+    # Sum over all k, and remove the terms j=k
+    # Don't subtract off the diagonal - this could be a NaN if distance to self is zero
+    np.fill_diagonal(competition_matrix, 0)
+    mysums = np.sum(competition_matrix, axis=1)
+    # Rather than computing that interior sum for each element, just compute the row sums,
+    # and then subtract the k=i term.
+    for i in range(len(pops)):
+        for j in range(len(pops)):
+            if j != i:
+                network[i][j] = network[i][j] * (mysums[j] - competition_matrix[j][i]) ** delta
+
+    np.fill_diagonal(network, 0)
+    return network
+
+
+def sum_populations_as_close_or_closer(sorted_pops, sorted_distance_row):
+    # Separating this operation out because it is common between a couple of migration models, and there is a little trickiness in appropriately
+    # handling cases where there are multiple destinations equidistant from the source node.
+    _is_instance(sorted_pops, np.ndarray, f"sorted_pops must be a NumPy array ({type(sorted_pops)=})")
+    _is_dtype(sorted_pops, np.number, f"sorted_pops must be a numeric array ({sorted_pops.dtype=})")
+    _has_values(sorted_pops >= 0, "sorted_pops must contain only non-negative values")
+    _is_instance(sorted_distance_row, np.ndarray, f"sorted_distance_row must be a NumPy array ({type(sorted_distance_row)=})")
+    _is_dtype(sorted_distance_row, np.number, f"sorted_distance_row must be a numeric array ({sorted_distance_row.dtype=})")
+    _has_values(sorted_distance_row >= 0, "sorted_distance_row must contain only non-negative values")
+    _has_shape(
+        sorted_pops,
+        sorted_distance_row.shape,
+        f"sorted_pops & sorted_distance_row must have same shape ({sorted_pops.shape=}, {sorted_distance_row.shape=})",
+    )
+    _has_values(np.diff(sorted_distance_row) >= 0, "sorted_distance_row must be sorted in ascending order")
+
+    # if all distances from the source node are unique, this is easy
+    cumulative_sorted_pops = np.cumsum(sorted_pops)
+
+    if len(np.unique(sorted_distance_row)) < len(sorted_distance_row):
```

Comment:
No test case passes this `if` statement, so the following code isn't ever executed or tested. We need a test case which will pass this check. Perhaps some synthetic data from a grid rather than the US cities data.